Source/JavaScriptCore/runtime/StringPrototype.cpp

```diff
@@ -1757,4 +1764,115 @@ JSC_DEFINE_HOST_FUNCTION(stringProtoFuncNormalize, (JSGlobalObject* globalObject
     RELEASE_AND_RETURN(scope, JSValue::encode(normalize(globalObject, string, form)));
 }
 
+static inline std::optional<unsigned> nonWellFormedIndex(const UChar* characters, unsigned length)
+{
+    for (unsigned index = 0; index < length; ++index) {
+        UChar character = characters[index];
+        if (!U16_IS_SURROGATE(character))
+            continue;
+
+        if (U16_IS_SURROGATE_TRAIL(character))
+            return index;
+
+        ASSERT(U16_IS_SURROGATE_LEAD(character));
+        if ((index + 1) == length)
+            return index;
+        UChar nextCharacter = characters[index + 1];
+
+        if (!U16_IS_SURROGATE(nextCharacter))
+            return index;
+
+        if (!U16_IS_SURROGATE_TRAIL(nextCharacter))
+            return index;
+
+        ++index; // Increment additionally.
+    }
+    return std::nullopt;
+}
+
+JSC_DEFINE_HOST_FUNCTION(stringProtoFuncIsWellFormed, (JSGlobalObject* globalObject, CallFrame* callFrame))
+{
+    VM& vm = globalObject->vm();
+    auto scope = DECLARE_THROW_SCOPE(vm);
+
+    JSValue thisValue = callFrame->thisValue();
+    if (!checkObjectCoercible(thisValue))
+        return throwVMTypeError(globalObject, scope);
+
+    if (thisValue.isString() && asString(thisValue)->is8Bit())
+        return JSValue::encode(jsBoolean(true));
+
+    String string = thisValue.toWTFString(globalObject);
+    RETURN_IF_EXCEPTION(scope, { });
+
+    if (string.is8Bit())
+        return JSValue::encode(jsBoolean(true)); // Latin-1 characters do not have surrogates.
+    return JSValue::encode(jsBoolean(!nonWellFormedIndex(string.characters16(), string.length())));
+}
+
+JSC_DEFINE_HOST_FUNCTION(stringProtoFuncToWellFormed, (JSGlobalObject* globalObject, CallFrame* callFrame))
+{
+    VM& vm = globalObject->vm();
+    auto scope = DECLARE_THROW_SCOPE(vm);
+
+    JSValue thisValue = callFrame->thisValue();
+    if (!checkObjectCoercible(thisValue))
+        return throwVMTypeError(globalObject, scope);
+
+    if (thisValue.isString() && asString(thisValue)->is8Bit())
+        return JSValue::encode(thisValue);
+
+    String string = thisValue.toWTFString(globalObject);
+    RETURN_IF_EXCEPTION(scope, { });
+
+    if (string.is8Bit())
+        return JSValue::encode(thisValue);
+
+    const UChar* characters = string.characters16();
+    unsigned length = string.length();
+    auto firstNonWellFormedIndex = nonWellFormedIndex(characters, length);
+    if (!firstNonWellFormedIndex)
+        return JSValue::encode(thisValue);
+
+    Vector<UChar> buffer;
+    buffer.reserveInitialCapacity(length);
+    buffer.append(characters, firstNonWellFormedIndex.value());
+    for (unsigned index = firstNonWellFormedIndex.value(); index < length; ++index) {
+        UChar character = characters[index];
+
+        if (!U16_IS_SURROGATE(character)) {
+            buffer.append(character);
+            continue;
+        }
+
+        if (U16_IS_SURROGATE_TRAIL(character)) {
+            buffer.append(replacementCharacter);
```

Comment:
Wow, it really surprised me that `replacementCharacter` is the name of a global. 😅 (Since it doesn't really sound like a _specific_ character...)